alphagov/cf-terraform
# [#103105066] Deploy Elastic Search server on AWS

[#103105066 Deploy Elastic Search server on AWS](https://www.pivotaltracker.com/story/show/103105066)
# What

In order to run Digital Marketplace we will need to have Elastic Search running too. This is to deploy an instance of Elastic Search using BOSH to AWS.
# Context

This story has a part of investigation to:
- test and chose the different bosh releases available
- find a way to add additional jobs to the official cloudfoundry release.

See [the card in pivotal for details](https://www.pivotaltracker.com/story/show/103105066)

Additionally, we took into account the cards:
- [#103010744 Normalise CF manifests between GCE and AWS](https://www.pivotaltracker.com/story/show/103010744)
- [#103313294 Deploy postgres server and broker](https://www.pivotaltracker.com/story/show/103313294)
# Scope of the PR

This card must be reviewed in this scope:
- I want to deploy one elasticsearch server. One unique data&master node running is enough.
- The deployment will be done using bosh.
- We want to deploy elasticsearch with the same manifest as Cloudfoundry, but keep using the [official manifest templates](https://github.com/cloudfoundry/cf-release/tree/master/templates)
- We should allow and make it easy support multiple Infrastructure platforms and add more additional jobs and services to the manifests.
# Implementation details and notes
- We added a new manifest generation script, and stop using terraform for generating a big stub file. See details in commit: 364ab5e1d146be4db171fc7b4d2a6be004c76875 
- We implement a deployment of elasticsearch in a3906904bad7a8738396cceaa0f94d4197aa1154. 

> Note: Although the [deployment file `manifests/templates/deployments/elasticsearch.yml`](https://github.com/alphagov/cf-terraform/blob/feature/103105066_elasticsearch/manifests/templates/deployments/elasticsearch.yml#L17--1) describes a complex cluster with 3 masters and 4 data nodes, later [in the stub `manifests/templates/stubs-aws/elasticsearch-stub.yml`](https://github.com/alphagov/cf-terraform/blob/feature/103105066_elasticsearch/manifests/templates/stubs-aws/elasticsearch-stub.yml#L12--1) we change the values to only deploy one master+data node. 
> 
> We did this way as an example of how to use the manifest generation using spiff. 
# How to test this PR

Start the environment:

```
make aws DEPLOY_ENV=<env>
```

A functional CF environment should be running, check it as described in #16.

Then test the ES node, by checking the cluster health and setting a value.

```
$ curl -XGET 'http://10.0.10.30:9200/_cluster/health?pretty=true'
{
  "cluster_name" : "elasticsearch_cluster",
  "status" : "green",
  "timed_out" : false,
  "number_of_nodes" : 1,
  "number_of_data_nodes" : 1,
  "active_primary_shards" : 0,
  "active_shards" : 0,
  "relocating_shards" : 0,
  "initializing_shards" : 0,
  "unassigned_shards" : 0,
  "number_of_pending_tasks" : 0
}

$ curl -XPUT 'http://10.0.10.30:9200/blog/user/dilbert' -d '{ "name" : "Dilbert Brown" }' ; echo
{"_index":"blog","_type":"user","_id":"dilbert","_version":1,"created":true}

$ curl 'http://10.0.10.30:9200/blog/user/dilbert' ; echo
{"_index":"blog","_type":"user","_id":"dilbert","_version":1,"found":true,"_source":{ "name" : "Dilbert Brown" }}

```
# who can review this?

anyone but @keymon or @saliceti
